### PR TITLE
C++: Promote `cpp/invalid-pointer-deref` out of experimental

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR moves the `cpp/invalid-pointer-deref` query out of experimental and into the `security-extended` suite.

Documentation friends: I've added a bunch of unnecessary whitespace in https://github.com/github/codeql/pull/14006/commits/e776178be52a09b3eb194a0b05bd91218dccfe67 that will ensure the relevant files show up in the diff. Once you're done reviewing the files I'll remove the whitespace again.

### CODE REVIEW

File: cpp/ql/src/Security/CWE/CWE-193/InvalidPointerDeref.ql [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    FinalFlow::PathNode source, FinalFlow::PathNode sink, int k, string kstr,
+    PointerArithmeticInstruction pai, string operation, Expr offset, DataFlow::Node n
+  where
+    k = min(int cand | hasFlowPath(source, sink, pai, operation, cand)) and
+    offset = pai.getRight().getUnconvertedResultExpression() and
+    n = source.getNode() and
+    if k = 0 then kstr = "" else kstr = " + " + k
+  select sink.getNode(), source, sink,
+    "This " + operation + " might be out of bounds, as the pointer might be equal to $@ + $@" + kstr +
+      ".", n, n.toString(), offset, offset.toString()
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-193/InvalidPointerDeref.ql]: Fixed in 51d51e34c23c3f7adb596880e612e80f91df98e0.

**Comment by mchammer01** [comment]
I'll review this for Docs.

**Review by mchammer01 — DISMISSED** [review]
LGTM ✨ 
Made a few comments and suggestions. Feel free to ignore the ones you don't agree with 😅

File: cpp/ql/src/Security/CWE/CWE-193/InvalidPointerDeref.ql [hunk 2]
```diff
@@ -0,0 +1,153 @@
+  /**
+   * @name Invalid pointer dereference
+   * @description Dereferencing a pointer that points past it allocation is undefined behavior
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-193/InvalidPointerDeref.ql:3]: That's indeed what I meant. Thank you! I'll try to think of a synonym of "to point" 🤔

**Comment by MathiasVP** [comment]
I'm not super concerned about the slowdown since Product has already told us that people are expecting a slowdown when running security-extended. But obviously the OOM is a problem (and not something we've seen in other runs, I think). So I'll investigate that 👀

**Comment by jketema** [comment]
> I'm not super concerned about the slowdown since Product has already told us that people are expecting a slowdown when running security-extended.

The slowdown is around 17% on average though, which is quite a lot.

**Comment by MathiasVP** [comment]
Indeed. I'll see if I can spot anything obvious we can do about it 🤔

**Comment by MathiasVP** [comment]
> It might help already if we can share the "bounded" predicates with the `cpp/overrun-write` query?

Yeah, that's a good idea.

**Comment by MathiasVP** [comment]
> > It might help already if we can share the "bounded" predicates with the `cpp/overrun-write` query?
> 
> Yeah, that's a good idea.

Fixed in 123e58767bd3ca18b7625ccede7494ac0ecf5c0a.

I've successfully run the security-extended suite locally with `--ram=3000` so I think this is simply another instance of OOM issues on the runner 😭.

**Comment by jketema** [comment]
Unfortunately no performance improvement after sharing the predicate 😞 Also still OOMs on OpenJDK, can the flag this up in the relevant internal discussion?

> I've successfully run the security-extended suite locally with --ram=3000 so I think this is simply another instance of OOM issues on the runner 😭.

This is a pretty bad test for this, as on the runner it's probably a 137 exit code, which means the container in which everything is run is getting killed.

**Comment by MathiasVP** [comment]
> Unfortunately no performance improvement after sharing the predicate 😞 Also still OOMs on OpenJDK. Even though it has OOMs it's also one of the projects that shows the biggest slowdown, so the clue might be in there?

Maybe. In terms of increased memory usage there's nothing immediately to point to. A bunch of extra work, sure (because we're running a new query), but nothing that would have increased memory pressure.

> This is a pretty bad test for this, as on the runner it's probably a 137 exit code, which means the container in which everything is run is getting killed.

I see. Do you have an idea for how I can test this better locally? 😭

**Comment by jketema** [comment]
> > This is a pretty bad test for this, as on the runner it's probably a 137 exit code, which means the container in which everything is run is getting killed.
> 
> I see. Do you have an idea for how I can test this better locally? 😭

Not really. This has been one of the challenges with these errors.

**Comment by MathiasVP** [comment]
Given the amount of optimization work I'm still doing on this PR I'm moving it back to a draft.

**Comment by MathiasVP** [comment]
The combination of join order fixes, a small amount of additional pruning, and (most importantly) disabling of field flow seems to have brought the evaluation time down to acceptable levels. So I think this PR is good to go! 🎉

**Review by jketema — APPROVED** [review]
LGTM

**Pull request merged**